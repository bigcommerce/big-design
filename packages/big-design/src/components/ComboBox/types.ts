import { Placement } from '@popperjs/core';
import React, { ComponentPropsWithoutRef, RefObject } from 'react';

import { InputProps } from '../Input';

interface BaseComboBox extends Omit<ComponentPropsWithoutRef<'input'>, 'children' | 'value'> {
  action?: ComboBoxAction;
  autoComplete?: string;
  autoWidth?: boolean;
  description?: React.ReactChild;
  disabled?: boolean;
  error?: InputProps['error'];
  filterable?: boolean;
  inputRef?: RefObject<HTMLInputElement> | React.Ref<HTMLInputElement>;
  label?: React.ReactChild;
  labelId?: string;
  maxHeight?: number;
  name?: string;
  optionsLoading?: boolean;
  placement?: Placement;
  positionFixed?: boolean;
  required?: boolean;
  onClose?(): void;
  onInputChange?(value: string): void;
  onOpen?(): void;
  onScrollToBottom?(): void;
}

export interface ComboBoxProps<T> extends BaseComboBox {
  options: Array<ComboBoxOption<T>> | Array<ComboBoxOptionGroup<T>>;
  value?: T;
  onOptionChange(value?: T, option?: ComboBoxOption<T>): void;
}

interface BaseItem extends Omit<ComponentPropsWithoutRef<'li'>, 'value'> {
  content: string;
  description?: string;
  disabled?: boolean;
  icon?: React.ReactElement;
  tooltip?: string;
}

export interface ComboBoxOption<T> extends BaseItem {
  value: T;
}

export interface ComboBoxAction extends BaseItem {
  actionType?: 'normal' | 'destructive';
  onActionClick(inputText: string | null): void;
}

export interface ComboBoxOptionGroup<T> {
  label: string;
  separated?: boolean;
  options: Array<ComboBoxOption<T>>;
}
