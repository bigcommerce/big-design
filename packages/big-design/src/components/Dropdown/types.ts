import { Placement } from '@popperjs/core';
import { ComponentPropsWithoutRef, ReactElement } from 'react';

export interface DropdownProps extends Omit<ComponentPropsWithoutRef<'ul'>, 'children'> {
  autoWidth?: boolean;
  disabled?: boolean;
  items: Array<DropdownItem | DropdownLinkItem> | DropdownItemGroup[];
  selectedItem?: DropdownItem;
  maxHeight?: number;
  placement?: Placement;
  positionFixed?: boolean;
  toggle: ReactElement<{ id?: string }>;
}

interface BaseItem extends Omit<ComponentPropsWithoutRef<'li'>, 'value'> {
  actionType?: 'normal' | 'destructive';
  content: string;
  description?: string;
  disabled?: boolean;
  icon?: ReactElement;
  tooltip?: string;
}

export interface DropdownItem extends BaseItem {
  hash?: string;
  type?: 'text';
  onItemClick(item: DropdownItem): void;
}

export interface DropdownLinkItem extends BaseItem {
  target?: HTMLAnchorElement['target'];
  type: 'link';
  url: string;
}

export interface DropdownItemGroup {
  label?: string;
  separated?: boolean;
  items: Array<DropdownItem | DropdownLinkItem>;
}
