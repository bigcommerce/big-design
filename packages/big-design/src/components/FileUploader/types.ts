import { MarginProps } from '../../helpers';
import { ButtonProps } from '../Button';

export interface Localization {
  upload: string;
  showLess: string;
  showMore: string;
  canNotUploadTheseFiles: string;
}

export interface ValidatorConfig {
  message?: string;
  type: string;
  validator: (file: File) => boolean;
}

export interface Action
  extends Omit<
    ButtonProps,
    'variant' | 'children' | 'iconOnly' | 'iconRight' | 'iconLeft' | 'isLoading' | keyof MarginProps
  > {
  label: string;
}
