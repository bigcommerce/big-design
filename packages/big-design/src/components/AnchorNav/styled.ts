import { theme as defaultTheme } from '@bigcommerce/big-design-theme';
import styled from 'styled-components';

import { BoxProps } from '../Box';

export const StyledAnchorNav = styled.div<BoxProps>`
display: block;
position relative;

.anchor-nav {
    &__list {
        position: sticky;
        inset-block-start: 0;
        width: 266px;
        float: inline-start;
        display:none;

        @media (min-width: ${({ theme }) => theme.breakpointValues.tablet}) {
            display: block;
            margin-inline-end: ${defaultTheme.spacing.xLarge};
        }

        & > ul {
            list-style-type: none;
            padding: 0;
            margin: 0;

            & > li {
                display:block;
                margin: 0;
                padding: 0;
                box-sizing: border-box;

                & > a {
                    color: ${defaultTheme.colors.primary};
                    text-decoration: none;
                    display: block;
                    padding: ${defaultTheme.spacing.small} ${defaultTheme.spacing.large};
                    border-inline-start: 3px solid transparent;

                    &:hover {
                        background-color: ${defaultTheme.colors.primary10};
                    }

                    &.active {
                        color: ${defaultTheme.colors.secondary70};
                        border-inline-start: 3px solid ${defaultTheme.colors.primary};

                        &:hover {
                            background-color: transparent;
                        }
                    }
                }
            }
        }
    
    }
    
    &__elements {
        display:flex;
        flex-direction: column;
        gap: ${defaultTheme.spacing.xLarge};
    }
}
`;

StyledAnchorNav.defaultProps = { theme: defaultTheme };
