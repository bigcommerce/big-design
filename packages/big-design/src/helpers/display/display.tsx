import { Breakpoints, breakpointsOrder, ThemeInterface } from '@bigcommerce/big-design-theme';
import { css, FlattenSimpleInterpolation } from 'styled-components';

import { DisplayOverload, DisplayProps } from './types';

export const withDisplay = () => css<DisplayProps>`
  ${({ display, theme }) => display && getDisplayStyles(display, theme, 'display')};
`;

const getDisplayStyles: DisplayOverload = (
  displayProp: any,
  theme: ThemeInterface,
  cssKey: any,
): FlattenSimpleInterpolation => {
  if (typeof displayProp === 'object') {
    return getResponsiveDisplay(displayProp, theme, cssKey);
  }

  if (typeof displayProp === 'string' || typeof displayProp === 'number') {
    return getSimpleDisplay(displayProp, cssKey);
  }

  return [];
};

const getSimpleDisplay = (
  displayProp: string | number,
  cssKey: string,
): FlattenSimpleInterpolation => css`
  ${cssKey}: ${displayProp}
`;

const getResponsiveDisplay: DisplayOverload = (
  displayProp: any,
  theme: ThemeInterface,
  cssKey: string,
): FlattenSimpleInterpolation[] => {
  const breakpointKeys = Object.keys(displayProp).sort(
    (firstBreakpoint, secondBreakpoint) =>
      // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
      breakpointsOrder.indexOf(firstBreakpoint as keyof Breakpoints) -
      // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
      breakpointsOrder.indexOf(secondBreakpoint as keyof Breakpoints),
  );

  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  return (breakpointKeys as Array<keyof Breakpoints>).map(
    (breakpointKey) => css`
      ${theme.breakpoints[breakpointKey]} {
        ${/* eslint-disable-next-line @typescript-eslint/no-unsafe-member-access */ ''}
        ${getSimpleDisplay(displayProp[breakpointKey], cssKey)}
      }
    `,
  );
};
