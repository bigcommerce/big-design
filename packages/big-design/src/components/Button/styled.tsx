import { addValues, theme as defaultTheme } from '@bigcommerce/big-design-theme';
import styled, { css } from 'styled-components';

import { MarginProps, withMargins } from '../../helpers';
import { withTransition } from '../../helpers/transitions';
import { Flex } from '../Flex';

import { ButtonProps } from './index';

export const StyledButton = styled.button<ButtonProps & MarginProps>`
  ${withTransition(['background-color', 'border-color', 'box-shadow', 'color'])}

  && {
    ${withMargins()};
  }

  align-items: center;
  appearance: none;
  border: ${({ theme }) => theme.border.box};
  border-radius: ${({ theme }) => theme.borderRadius.normal};
  color: ${({ theme }) => theme.colors.white};
  cursor: pointer;
  display: inline-flex;
  flex: none;
  font-size: ${({ theme }) => theme.typography.fontSize.medium};
  font-weight: ${({ theme }) => theme.typography.fontWeight.regular};
  height: ${({ theme }) => addValues(theme.spacing.xxLarge, theme.spacing.xxSmall)};
  justify-content: center;
  line-height: ${({ theme }) => theme.lineHeight.xLarge};
  outline: none;
  padding: ${({ theme }) => `0 ${theme.spacing.medium}`};
  position: relative;
  text-align: center;
  text-decoration: none;
  user-select: none;
  vertical-align: middle;
  white-space: nowrap;
  width: ${({ mobileWidth }) => (mobileWidth === 'auto' ? 'auto' : '100%')};

  &:focus {
    outline: none;
  }

  &[disabled] {
    border-color: ${({ theme }) => theme.colors.secondary30};
    pointer-events: none;
  }

  & + .bd-button {
    margin-top: ${({ mobileWidth, theme }) => mobileWidth === '100%' && theme.spacing.xSmall};
    margin-left: ${({ mobileWidth, theme }) => mobileWidth === 'auto' && theme.spacing.xSmall};

    ${({ theme }) => theme.breakpoints.tablet} {
      margin-top: ${({ theme }) => theme.spacing.none};
      margin-left: ${({ theme }) => theme.spacing.xSmall};
    }
  }

  ${({ theme }) => theme.breakpoints.tablet} {
    width: auto;

    ${({ iconOnly: icon, theme }) =>
      icon &&
      css`
        padding: 0;
        min-width: ${addValues(theme.spacing.xxLarge, theme.spacing.xxSmall)};
      `};
  }

  ${({ iconLeft, theme }) =>
    iconLeft &&
    css`
      padding-left: ${theme.spacing.xSmall};
    `};

  ${({ iconRight, theme }) =>
    iconRight &&
    css`
      padding-right: ${theme.spacing.xSmall};
    `};

  ${(props) => getButtonStyles(props)}
`;

export const ContentWrapper = styled.span.attrs<Record<string, unknown>, { isLoading?: boolean }>(
  {},
)`
  align-content: center;
  align-items: center;
  display: inline-grid;
  grid-auto-flow: column;
  grid-gap: ${({ theme }) => theme.spacing.xSmall};

  ${({ isLoading }) =>
    isLoading &&
    css`
      visibility: hidden;
    `};
`;

export const LoadingSpinnerWrapper = styled(Flex)`
  position: absolute;
`;

/**
 * These can be generated dynamically but I'm leaning towards being extra
 * explicit and being able to handle corner cases and changes from design easily
 */
const ButtonPrimary = css<ButtonProps>`
  background-color: ${({ theme }) => theme.colors.primary};
  border-color: ${({ theme }) => theme.colors.primary};
  font-weight: ${({ theme }) => theme.typography.fontWeight.semiBold};

  &:active {
    background-color: ${({ theme }) => theme.colors.primary60};
  }

  &:focus {
    box-shadow: ${({ theme }) => `0 0 0 ${theme.spacing.xxSmall} ${theme.colors.primary20}`};
  }

  &:hover:not(:active) {
    background-color: ${({ theme }) => theme.colors.primary50};
  }

  &[disabled] {
    background-color: ${({ theme }) => theme.colors.secondary30};
  }
`;

const ButtonPrimaryDestructive = css<ButtonProps>`
  background-color: ${({ theme }) => theme.colors.danger};
  border-color: ${({ theme }) => theme.colors.danger};
  font-weight: ${({ theme }) => theme.typography.fontWeight.semiBold};

  &:active {
    background-color: ${({ theme }) => theme.colors.danger60};
  }

  &:focus {
    box-shadow: ${({ theme }) => `0 0 0 ${theme.spacing.xxSmall} ${theme.colors.danger20}`};
  }

  &:hover:not(:active) {
    background-color: ${({ theme }) => theme.colors.danger50};
  }

  &[disabled] {
    background-color: ${({ theme }) => theme.colors.secondary30};
  }
`;

const ButtonSecondary = css<ButtonProps>`
  background-color: transparent;
  border-color: ${({ theme }) => theme.colors.primary};
  color: ${({ theme }) => theme.colors.primary};

  &:active {
    background-color: ${({ theme }) => theme.colors.primary20};
  }

  &:focus {
    box-shadow: ${({ theme }) => `0 0 0 ${theme.spacing.xxSmall} ${theme.colors.primary20}`};
  }

  &:hover:not(:active) {
    background-color: ${({ theme }) => theme.colors.primary10};
  }

  &[disabled] {
    color: ${({ theme }) => theme.colors.secondary30};
  }
`;

const ButtonSecondaryDestructive = css<ButtonProps>`
  background-color: transparent;
  border-color: ${({ theme }) => theme.colors.danger};
  color: ${({ theme }) => theme.colors.danger};

  &:active {
    background-color: ${({ theme }) => theme.colors.danger20};
  }

  &:focus {
    box-shadow: ${({ theme }) => `0 0 0 ${theme.spacing.xxSmall} ${theme.colors.danger20}`};
  }

  &:hover:not(:active) {
    background-color: ${({ theme }) => theme.colors.danger10};
  }

  &[disabled] {
    color: ${({ theme }) => theme.colors.secondary30};
  }
`;

const ButtonSubtle = css<ButtonProps>`
  background-color: transparent;
  border-color: transparent;
  color: ${({ theme }) => theme.colors.primary};

  &:active {
    background-color: ${({ theme }) => theme.colors.primary20};
  }

  &:focus {
    box-shadow: ${({ theme }) => `0 0 0 ${theme.spacing.xxSmall} ${theme.colors.primary20}`};
  }

  &:hover:not(:active) {
    background-color: ${({ theme }) => theme.colors.primary10};
  }

  &[disabled] {
    border-color: transparent;
    color: ${({ theme }) => theme.colors.secondary30};
  }
`;

const ButtonSubtleDestructive = css<ButtonProps>`
  background-color: transparent;
  border-color: transparent;
  color: ${({ theme }) => theme.colors.danger};

  &:active {
    background-color: ${({ theme }) => theme.colors.danger20};
  }

  &:focus {
    box-shadow: ${({ theme }) => `0 0 0 ${theme.spacing.xxSmall} ${theme.colors.danger20}`};
  }

  &:hover:not(:active) {
    background-color: ${({ theme }) => theme.colors.danger10};
  }

  &[disabled] {
    border-color: transparent;
    color: ${({ theme }) => theme.colors.secondary30};
  }
`;

const ButtonUtility = css<ButtonProps>`
  background-color: transparent;
  border-color: transparent;
  color: ${({ theme }) => theme.colors.secondary60};
  padding-inline: ${({ theme }) => theme.spacing.xxSmall};
  min-width: ${({ theme }) => addValues(theme.spacing.xLarge, theme.spacing.small)};

  &:active {
    background-color: ${({ theme }) => theme.colors.primary20};
    color: ${({ theme }) => theme.colors.primary60};
  }

  &:focus:not(:active) {
    box-shadow: ${({ theme }) => `0 0 0 ${theme.spacing.xxSmall} ${theme.colors.secondary10}`};
    color: ${({ theme }) => theme.colors.primary50};
  }

  &:hover:not(:active) {
    background-color: ${({ theme }) => theme.colors.primary10};
    color: ${({ theme }) => theme.colors.primary50};
  }

  &[disabled] {
    border-color: transparent;
    color: ${({ theme }) => theme.colors.secondary50};
  }
`;

const ButtonUtilityDestructive = css<ButtonProps>`
  background-color: transparent;
  border-color: transparent;
  color: ${({ theme }) => theme.colors.secondary60};

  &:active {
    background-color: ${({ theme }) => theme.colors.danger20};
    color: ${({ theme }) => theme.colors.danger60};
  }

  &:focus:not(:active) {
    box-shadow: ${({ theme }) => `0 0 0 ${theme.spacing.xxSmall} ${theme.colors.danger20}`};
    color: ${({ theme }) => theme.colors.danger40};
  }

  &:hover:not(:active) {
    background-color: ${({ theme }) => theme.colors.danger10};
    color: ${({ theme }) => theme.colors.danger50};
  }

  &[disabled] {
    border-color: transparent;
    color: ${({ theme }) => theme.colors.secondary50};
  }
`;

function getButtonStyles(props: ButtonProps) {
  const { actionType, variant } = props;

  switch (variant) {
    case 'primary':
      return actionType === 'destructive' ? ButtonPrimaryDestructive : ButtonPrimary;

    case 'secondary':
      return actionType === 'destructive' ? ButtonSecondaryDestructive : ButtonSecondary;

    case 'subtle':
      return actionType === 'destructive' ? ButtonSubtleDestructive : ButtonSubtle;

    case 'utility':
      return actionType === 'destructive' ? ButtonUtilityDestructive : ButtonUtility;
  }
}

StyledButton.defaultProps = { theme: defaultTheme };
ContentWrapper.defaultProps = { theme: defaultTheme };
LoadingSpinnerWrapper.defaultProps = { theme: defaultTheme };
