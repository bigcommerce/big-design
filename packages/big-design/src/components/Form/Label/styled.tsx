import { theme as defaultTheme } from '@bigcommerce/big-design-theme';
import styled, { DefaultTheme, StyledComponent } from 'styled-components';

import { StyleableH4 } from '../../Typography/private';
import { HeadingProps } from '../../Typography/types';

import { type FormControlLabelProps } from './Label';

interface StyledLabelArgument extends Omit<FormControlLabelProps, 'localization'> {
  theme: DefaultTheme;
}

export const StyledLabel = styled<
  StyledComponent<'label' | 'h4', DefaultTheme, Partial<HeadingProps>> & FormControlLabelProps
>(StyleableH4).attrs({
  as: 'label',
})<FormControlLabelProps>`
  cursor: pointer;
  display: inline-block;
  margin-bottom: ${({ theme }: StyledLabelArgument) => theme.spacing.xxSmall};
`;

StyledLabel.defaultProps = { theme: defaultTheme };
