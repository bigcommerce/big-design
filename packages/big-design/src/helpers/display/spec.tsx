import { theme as defaultTheme } from '@bigcommerce/big-design-theme';
import 'jest-styled-components';
import { render, screen } from '@testing-library/react';
import React from 'react';
import styled from 'styled-components';

import { withDisplay } from './display';
import { DisplayProps } from './types';

const TestComponent = styled.div<DisplayProps>`
  ${withDisplay()};
`;

TestComponent.defaultProps = { theme: defaultTheme };

test('display', () => {
  render(<TestComponent data-testid="display-test" display="inline-block" />);

  expect(screen.getByTestId('display-test')).toHaveStyle('display: inline-block');
});

test('responsive display', () => {
  render(
    <TestComponent data-testid="display-responsive" display={{ mobile: 'none', tablet: 'flex' }} />,
  );

  expect(screen.getByTestId('display-responsive')).toMatchSnapshot();
});
