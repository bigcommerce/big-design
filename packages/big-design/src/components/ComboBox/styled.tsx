import { ArrowDropDownIcon } from '@bigcommerce/big-design-icons';
import { theme as defaultTheme } from '@bigcommerce/big-design-theme';
import { hideVisually } from 'polished';
import styled, { css } from 'styled-components';

import { StyleableButton } from '../Button/private';

export const StyledStatusMessage = styled.div`
  ${hideVisually()}
`;

export const StyledDropdownIcon = styled(ArrowDropDownIcon)<{ disabled?: boolean }>`
  ${({ disabled }) =>
    !disabled &&
    css`
      :hover {
        cursor: pointer;
      }
    `}
`;

export const StyledInputContainer = styled.div`
  input[readonly] {
    cursor: pointer;
  }
`;

export const DropdownButton = styled(StyleableButton)`
  color: ${({ theme }) => theme.colors.secondary60};
  height: auto;
  padding: 0;
`;

StyledStatusMessage.defaultProps = { theme: defaultTheme };
StyledDropdownIcon.defaultProps = { theme: defaultTheme };
StyledInputContainer.defaultProps = { theme: defaultTheme };
DropdownButton.defaultProps = { theme: defaultTheme };
