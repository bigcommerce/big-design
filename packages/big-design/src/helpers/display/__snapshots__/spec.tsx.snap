// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`responsive display 1`] = `
@media (min-width:0px) {
  .c0 {
    display: none;
  }
}

@media (min-width:720px) {
  .c0 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
  }
}

<div
  class="c0"
  data-testid="display-responsive"
  display="[object Object]"
/>
`;
